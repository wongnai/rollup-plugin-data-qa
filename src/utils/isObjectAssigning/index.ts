export default function isObjectAssigning(node: Record<string, any>) {
	return node.callee?.object?.name === 'Object' && node.callee?.property?.name === 'assign'
}
