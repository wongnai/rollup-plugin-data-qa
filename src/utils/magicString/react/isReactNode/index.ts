export default function isReactNode(node: Record<string, any>) {
	const isReactCreateElement =
		node?.callee?.object?.name === 'React' && node?.callee?.property?.name === 'createElement'

	const isJsxAndNotFragment =
		(node?.callee?.name === 'jsxDEV' ||
			node?.callee?.name === 'jsx' ||
			node?.callee?.name === 'jsxs') &&
		node?.arguments?.[0]?.name !== 'Fragment'

	return isReactCreateElement || isJsxAndNotFragment
}
