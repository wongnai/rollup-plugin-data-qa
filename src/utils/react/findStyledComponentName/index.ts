export default function findStyledComponentName(node: Record<string, any>): string {
	if (node.type === 'VariableDeclarator' && node.id) {
		return node.id.name
	}

	return ''
}
