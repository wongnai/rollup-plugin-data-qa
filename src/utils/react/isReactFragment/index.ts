import { FRAGMENT_NAMES } from 'pluginConstants'

export default function isReactFragment(node: Record<string, any>) {
	return FRAGMENT_NAMES.includes(node?.arguments?.[0]?.name)
}
