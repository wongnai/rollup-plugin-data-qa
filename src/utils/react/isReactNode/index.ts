import { JSX_CALLEE_NAMES } from 'pluginConstants'
import isReactFragment from 'utils/react/isReactFragment'

export default function isReactNode(node: Record<string, any>) {
	const isReactCreateElement =
		node?.callee?.object?.name === 'React' && node?.callee?.property?.name === 'createElement'

	const isJsxAndNotFragment =
		JSX_CALLEE_NAMES.includes(node?.callee?.name) && !isReactFragment(node)

	return isReactCreateElement || isJsxAndNotFragment
}
