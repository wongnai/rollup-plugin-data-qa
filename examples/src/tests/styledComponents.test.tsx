import styled from 'styled-components'
import { render } from '@testing-library/react'
import { screen } from '../utils/testUtils'
import { useMemo } from 'react'

describe('styled components', () => {
	it('01 - styled component in a div', () => {
		const StyledA = styled.div`
			color: red;
		`

		const ComponentA = () => {
			return (
				<div>
					<StyledA>xxx</StyledA>
				</div>
			)
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('styled-a')).toBeVisible()
		expect(screen.getByDataQa('component-a')).toBeVisible()
	})

	it('02 - returning styled component', () => {
		const StyledA = styled.div`
			color: red;
		`

		const ComponentA = () => {
			return <StyledA>xxx</StyledA>
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
		expect(screen.queryByDataQa('styled-a')).not.toBeInTheDocument()
	})

	it('03 - returning styled component but with useMemo above', () => {
		const StyledA = styled.div`
			color: red;
		`

		const ComponentA = () => {
			const B = useMemo(() => <div>xxx</div>, [])

			return <StyledA>{B}</StyledA>
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
		expect(screen.getByDataQa('b')).toBeVisible()
		expect(screen.queryByDataQa('styled-a')).not.toBeInTheDocument()
	})
})
