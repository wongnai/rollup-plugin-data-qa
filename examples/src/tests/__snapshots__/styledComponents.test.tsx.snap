// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`styled components > 01 - styled component in a div 1`] = `
.c0 {
  color: red;
}

<div
  data-qa="component-a"
>
  <div
    class="c0"
    data-qa="styled-a"
  >
    xxx
  </div>
</div>
`;

exports[`styled components > 02 - returning styled component 1`] = `
.c0 {
  color: red;
}

<div
  class="c0"
  data-qa="component-a"
>
  xxx
</div>
`;

exports[`styled components > 03 - returning styled component but with useMemo above 1`] = `
.c0 {
  color: red;
}

<div
  class="c0"
  data-qa="component-a"
>
  <div>
    xxx
  </div>
</div>
`;
